cpp/include/cudf/detail/gather.cuh

```diff
@@ -46,56 +46,104 @@ struct bounds_checker {
+ * @brief For each mask, assigns each bit to a value obtained by calling the
+ *        validity selector.
+ *
+ * @note If any mask is `nullptr`, that mask will be ignored.
+ *
+ * @param get_validity    Function used to obtain validity for each bit.
+ *                        Signature: `bool(size_type mask_idx, size_type bit_idx)`
+ * @param masks           Masks for which bits will be obtained and assigned.
+ * @param mask_count      The number of `masks`.
+ * @param mask_num_bits   The number of bits to assign for each mask. If this
+ *                        number is smaller than the total number of bits, the
+ *                        remaining bits will be set to `0` / `false`.
+ * @param validity_counts Used to obtain the total number of valid bits for each
+ *                        mask.
+ *
+ * @tparam ValiditySelector bool(size_type mask_idx, size_type bit_idx);
+ *---------------------------------------------------------------------------**/
+template<typename ValiditySelector>
+__device__ void select_bitmask(ValiditySelector get_validity,
+                               bitmask_type* masks[],
+                               size_type mask_count,
+                               size_type mask_num_bits,
+                               size_type* validity_counts)
+{
+  for (size_type mask_idx = 0; mask_idx < mask_count; mask_idx++) {
+    auto const mask = masks[mask_idx];
+    if (mask == nullptr) {
+      continue;
+    }
 
+    auto bit_idx_offset = blockIdx.x * blockDim.x;
+    auto validity_count_aggregate = static_cast<size_type>(0);
 
-template <bool ignore_out_of_bounds, typename MapIterator>
-__global__ void gather_bitmask_kernel(table_device_view source_table,
-                                      MapIterator gather_map,
-                                      bitmask_type * masks[],
-                                      size_type destination_table_num_rows,
-                                      size_type* valid_counts) {
-
-  for (size_type i = 0; i < source_table.num_columns(); i++) {
-
-    column_device_view source_col = source_table.column(i);
-
-    if (masks[i] != nullptr) {
-      size_type destination_row_base = blockIdx.x * blockDim.x;
-      cudf::size_type valid_count_accumulate = 0;
-
-      while (destination_row_base < destination_table_num_rows) {
-        size_type destination_row = destination_row_base + threadIdx.x;
+    while (bit_idx_offset < mask_num_bits) {
+      auto const bit_idx = bit_idx_offset + threadIdx.x;
+      auto const bit_active = bit_idx < mask_num_bits;
+      auto const bit_is_valid = bit_active && get_validity(mask_idx, bit_idx);
+      auto const warp_validity = __ballot_sync(0xffffffff, bit_is_valid);
+      auto const validity_idx = word_index(bit_idx);
 
-        const bool thread_active = destination_row < destination_table_num_rows;
-        size_type source_row =
-          thread_active ? gather_map[destination_row] : 0;
+      if (threadIdx.x % warp_size == 0) {
```

Comment:
```suggestion
      if (threadIdx.x % warp_size == 0 && bit_active) {
```
If `threadIdx.x >= warp_size`, wouldn't it be possible for `bit_idx >= mask_num_bits`?

Reply:
Good catch! Thanks!